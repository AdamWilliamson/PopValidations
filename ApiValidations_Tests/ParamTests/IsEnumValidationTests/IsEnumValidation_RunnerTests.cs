using ApprovalTests;
using FluentAssertions;
using ApiValidations_Tests.TestHelpers;
using ApiValidations;
using ApiValidations_Tests.ValidationsTests.IsEmptyValidationTests.IsEmptyValidationTests;

namespace ApiValidations_Tests.ValidationsTests.IsEnumToValidationTests;

public enum IsEnumTestEnum
{
    First = 10,
    Second = 20
}

public class EnumApi
{
    public void IsEnum(IsEnumTestEnum param1, string param2, int param3) { }
}

public class IsEnum_TestingValidator : ApiValidator<EnumApi>
{
    public IsEnum_TestingValidator()
    {
        DescribeFunc(x => x.IsEnum(
            Param.Is<IsEnumTestEnum>().IsEnum(typeof(IsEnumTestEnum)),
            Param.Is<string>().IsEnum(typeof(IsEnumTestEnum)),
            Param.Is<int>().IsEnum(typeof(IsEnumTestEnum))
        ));
    }
}

public class IsEnumValidation_RunnerTests
{
    [Fact]
    public void WhenDescribing_ItReturnsTheValidation()
    {
        // Arrange
        var runner = ValidationRunnerHelper.BasicRunnerSetup(new IsEnum_TestingValidator());

        // Act
        var description = runner.Describe();

        // Assert
        description.Results.Should().HaveCount(3);
        description.Results.Should().HaveCount(ValidatableHelper.GetValidatableCount<EmptyApi>(ValidatableType.NoExceptions));
        Approvals.VerifyJson(JsonConverter.ToJson(description));
    }
}
