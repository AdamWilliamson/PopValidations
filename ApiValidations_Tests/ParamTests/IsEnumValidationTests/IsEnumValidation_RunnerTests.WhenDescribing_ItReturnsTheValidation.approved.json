{
    "results": [
        {
            "property": "IsEnum(IsEnumTestEnum,String,Int32):Void::(param1,0,IsEnumTestEnum)",
            "outcomes": [
                {
                    "validator": "IsEnumValidation",
                    "message": "Must be one of 'First,Second' or '10,20'.",
                    "values": [
                        {
                            "key": "enumNames",
                            "value": "First,Second"
                        },
                        {
                            "key": "enumValues",
                            "value": "10,20"
                        },
                        {
                            "key": "fieldType",
                            "value": "enum"
                        }
                    ]
                }
            ],
            "validationGroups": []
        },
        {
            "property": "IsEnum(IsEnumTestEnum,String,Int32):Void::(param2,1,String)",
            "outcomes": [
                {
                    "validator": "IsEnumValidation",
                    "message": "Must be one of 'First,Second' or '10,20'.",
                    "values": [
                        {
                            "key": "enumNames",
                            "value": "First,Second"
                        },
                        {
                            "key": "enumValues",
                            "value": "10,20"
                        },
                        {
                            "key": "fieldType",
                            "value": "string"
                        }
                    ]
                }
            ],
            "validationGroups": []
        },
        {
            "property": "IsEnum(IsEnumTestEnum,String,Int32):Void::(param3,2,Int32)",
            "outcomes": [
                {
                    "validator": "IsEnumValidation",
                    "message": "Must be one of 'First,Second' or '10,20'.",
                    "values": [
                        {
                            "key": "enumNames",
                            "value": "First,Second"
                        },
                        {
                            "key": "enumValues",
                            "value": "10,20"
                        },
                        {
                            "key": "fieldType",
                            "value": "numeric"
                        }
                    ]
                }
            ],
            "validationGroups": []
        }
    ]
}